### ISSUE TITLE
Adding "Authenticate when any URI is accessed" in Liberty

### ISSUE THREAD

**Issue by NottyCode** [OpenLiberty/open-liberty#14634]
Some applications may not have Java EE security roles, but still need users accessing to be authenticated. This has been requested by IBM customers and is available in other runtimes. 

When ready, add links to the Upcoming Feature Overview document and Feature Test Summary issue:
- UFO:
- FTS:

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## List of Steps to complete or get approvals / sign-offs for Onboarding to the Liberty release (GM date)

Instructions:
- Do the actions below and mark them complete in the checklist when they are done.
- Make sure all feature readiness approvers put the appropriate tag on the epic to indicate their approval.

## **Design**
Before Development Starts or 8 weeks before Onboarding
- [ ] POC Design / UFO Review Scheduled (David Chang) or N/A.
- [ ] POC Design / UFO Reviewed (Feature Owner) or N/A.
- [ ] Complete any follow-ons from the POC Review.
- [ ] Design / UFO Approval ([Alasdair Nottingham](https://github.com/NottyCode)) or N/A.
- [ ] No Design / No UFO Approval ([Arthur De Magalhaes](https://github.com/arthurdm) - cloud / [Alasdair Nottingham](https://github.com/NottyCode) - server) or N/A.
- [ ] SVT Requirements identified. (Epic owner / Feature owner with SVT focal point)
- [ ] ID Requirements identified. (Epic owner / Feature owner with ID focal point)
- [ ] Create a child task of this epic entitled "FAT Approval Test Summary". Add the link in above.

## **Legal**
3 weeks before Onboarding
- [ ] Identify all open source libraries that are changing or are new. Work with Legal Release Services (Cass Tucker or Release PM) to get open source cleared and approved. Or N/A. (Epic Owner).   New or changed open source impacts license and Certificate of Originality.

## **Translation**
3 weeks before Onboarding
- [ ] All new or changed PII messages are checked into the integration branch, before the last translation shipment out. (Epic Owner)

## **Feature Complete**
2 weeks before Onboarding
- [ ] Implementation complete. (Epic owner / Feature owner)
- [ ] All function tests complete. Ready for FAT Approval. (Epic owner / Feature owner)
- [ ] Review all known issues for Stop Ship. (Epic owner / Feature owner / PM)

## **Focal Point Approvals**
2 to 1 week before Onboarding
#### You **MUST** have the Design Approved or No Design Approved label before requesting focal point approvals.

All features (both "Design Approved" and "No Design Approved")
- [ ] **FAT** - ([Adam Yoho](https://github.com/ayoho), [Martin Holder](https://github.com/mhldr), or [Dave Waddling](https://github.com/dave-waddling)). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [ ] **Demo** - ([Tom Evans](https://github.com/tevans78) or [Chuck Bridgham](https://github.com/cbridgha)). Demo is scheduled for an upcoming EOI. Approver adds label focalApproved:demo to the Epic in Github.
- [ ] **Globalization** ([Sam Wong](https://github.com/samwatibm) - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.

"Design Approved" features
- [ ] **Accessibility** - ([Steven Zvonek](https://github.com/steven1046)). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [ ] **ID** - ([Kareen Deen](https://github.com/chirp1)). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
- [ ] **Performance** - ([Jared Anderson](https://github.com/jhanders34)). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [ ] **Serviceability** - ([Don Bourne](https://github.com/donbourne)). Serviceability has been addressed.
- [ ] **STE** - (Swati Kasundra). STE chart deck is complete or N/A . Approver adds label focalApproved:ste to the Epic in Github.
- [ ] **SVT** - ([Brian Hanczaryk](https://github.com/hanczaryk) - APS). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github.

## **Ready for GA**
1 week before Onboarding
- [ ] No Stop Ship issues for the feature. (Epic owner / Feature owner / Release PM)
- [ ] Ship Readiness Review and Release Notes completed (Epic owner / Feature owner / Release PM)
- [ ] Github Epic and Epic's issues are closed / complete. All PRs are committed to the master branch. (Epic owner / Feature owner / Backlog Subtribe PM)

## **Other deliverbles**
- [ ] **OL Guides** - ([Yee-Kang Chang](https://github.com/yeekangc)). Assessment for OL Guides is complete or N/A.
- [ ] **WDT** - (Leonard Theivendra). WDT work complete or N/A.
- [ ] **Blog** - ([Laura Cowen](https://github.com/lauracowen)) Blog article writeup (Epic owner / Feature owner / Laura Cowen)

**Comment by scottkurz** [comment]
## UFO Socialization Details
Date: Feb. 27, 2023
Presenter: Jesse Van Hill
Link to UFO meeting recording:  https://ibm.box.com/s/3u26fvz63ob9db68gb0miuxhiz8vsqvg

## Notes:

### Investigate:

There was one broad question which was to understand how this new setting would or wouldn't affect "system apps", feature function that is not packaged and deployed like an application (WAR/EAR etc.) but internally plugs into things like the webcontainer to leverage it like an app does, to provide feature implementation-provided endpoints:  e.g. feature function implemented with the help of the REST Handler.  

Some examples listed were:

* MP Metrics - Has its own security configuration : https://openliberty.io/docs/latest/reference/feature/mpMetrics-4.0.html#_configure_microprofile_metrics_security and means for disabling: https://openliberty.io/docs/latest/reference/feature/mpMetrics-4.0.html#_remove_microprofile_metrics_security
* Batch - https://www.ibm.com/docs/en/was-liberty/nd?topic=liberty-securing-batch-environment - Uses REST Handler, keys off of the presence/absence of the app security feature but not other Java EE / Jakarta security config like `@ServletSecurity` provides)
* MP Health - Which I believe is typically (always?) unauthenticated

We presumably wouldn't want to change/break the security-related workflow of these features by enabling something for normal, user web apps globally.

Team should investigate and decide whether any concerns can be accommodated in a straightforward way or if more design socialization should be done?

### Minor Notes

1.  slide 11 - End User Overview  - Is it OK to list '403' as the error code or is it better to indicate some mix of '401' vs. '403'?

`JVH - I am going to leave this as is. If they provided (invalid) credentials on the request, they would immediately get a 403. If they did not, they would get a 401 and then upon sending the (invalid) credentials they would get a 403. The 401 would be part of the "Prompt Login" diagram to the left of "Authenticate"`


2. slide 16 - (config) -  Maybe rephrase:  "Add the following elements:" => "Add this attribute to the following elements:"  (_not part of recording, I just noticed it myself here_)

`JVH - I rephrased to "Add these new attributes to the following elements:".`

3. slide 16 (and more generally) - "protectAllURIs vs protectAllURLs" ?  TWAS setting mentions "URI", so some precedence there.  Newer doc generally prefers "URI".   Doesn't seem like there's a meanginful non-URL URI so maybe "URL" is preferable (since it is more specific)?    Discussed but left solution to implementing team, to socialize if/as needed

`JVH - Decided to match the url-pattern in auth-constraints in the web.xml, so we named it protectAllURLs.`

4. slide 31 - serviceability - One or both of the Tr.audit messages need to be removed?

`JVH - I deleted this slide. It should have been removed after initial reviews with my team.`

5. Feature Design (slides 9-10 maybe?) - Mention the fact that app security feature needs to be enabled for any of this to apply.   E.g. enabling the global protectAllURIs doesn't cause auth to be required if the app security feature is not configured.
( **NOTE:** whatever is said should take into account the "system app" discussion above, as necessary)

`JVH - I added a statement to slide 9.`

### Other 

Then there were two discussions I can't easily summarize:

1. Investigate the behavior for SpringBoot applications, including each of WAR and JAR app packaging types

`JVH - The security team talked it over with @tjwatson and he indicated that spring boot apps configure their own security and don't rely on our web app security. Therefore; we decided that Spring Boot applications should remain unaffected by this change. Therefore there will be no need to add a an attribute in sprintBootApplication as we will handle it internally. Added a new slide (10) that covers this.`

2. Consider behavior for HTTP (as opposed to HTTPS) requests... can/should this force them to use HTTPS?   It sounded like there are differences here between Liberty & traditional WebSphere.  It sounds like Liberty has options to guarantee a confidential transport whereas traditional does not.   The discussion also expanded to touch on the question of can we detect the passing of credentials over an HTTP request.  

`JVH - When protectAllURLs is enabled, Liberty will treat all unprotected endpoints as if the transport-guarantee has been set to confidential. This will ensure that credentials are not passed in the clear should Liberty prompt for them. This will not protect against requests that preemptively send the credentials to the HTTP endpoint. I have added this to slide 10.`

**Comment by malincoln** [comment]
Per @arkarkala, 4-6 PW, setting size to L
